File: Number of Subsequences That Satisfy the Given Sum Condition/Number of Subsequences That Satisfy the Given Sum Condition.cpp
// Solution for Number of Subsequences That Satisfy the Given Sum Condition in CPP

class Solution {
public:
    const int MOD = 1e9 + 7;

    int numSubseq(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end()); // Sort to easily find min & max

        // Precompute powers of 2 modulo MOD
        vector<int> powerOfTwo(n, 1);
        for (int i = 1; i < n; i++) {
            powerOfTwo[i] = (2LL * powerOfTwo[i - 1]) % MOD;
        }

        int left = 0, right = n - 1;
        int totalValid = 0;

        // Two-pointer approach to check all valid pairs (min, max)
        while (left <= right) {
            if (nums[left] + nums[right] <= target) {
                // All subsequences formed between left and right are valid
                totalValid = (totalValid + powerOfTwo[right - left]) % MOD;
                left++;
            } else {
                // Reduce the max element to try to make sum smaller
                right--;
            }
        }

        return totalValid;
    }
};

File: Number of Subsequences That Satisfy the Given Sum Condition/Number of Subsequences That Satisfy the Given Sum Condition.py
# Solution for Number of Subsequences That Satisfy the Given Sum Condition in PY

class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        MOD = 10**9 + 7
        nums.sort()
        n = len(nums)

        power_of_two = [1] * n
        for i in range(1, n):
            power_of_two[i] = (power_of_two[i - 1] * 2) % MOD

        left, right = 0, n - 1
        total_valid = 0

        while left <= right:
            if nums[left] + nums[right] <= target:
                total_valid = (total_valid + power_of_two[right - left]) % MOD
                left += 1
            else:
                right -= 1

        return total_valid
